diff --git a/apt-pkg/aptconfiguration.cc b/apt-pkg/aptconfiguration.cc
index 9982759..9b4b46d 100644
--- a/apt-pkg/aptconfiguration.cc
+++ b/apt-pkg/aptconfiguration.cc
@@ -31,6 +31,12 @@
 
 #include <apti18n.h>
 									/*}}}*/
+
+#ifdef setlocale
+// Someone in libintl world decided it was a good idea to define a "setlocale" macro.
+#undef setlocale
+#endif
+
 namespace APT {
 // getCompressionTypes - Return Vector of usable compressiontypes	/*{{{*/
 // ---------------------------------------------------------------------
diff --git a/buildlib/config.h.in b/buildlib/config.h.in
index 6b72fb3..c6627d5 100644
--- a/buildlib/config.h.in
+++ b/buildlib/config.h.in
@@ -43,6 +43,15 @@
 /* The mail address to reach upstream */
 #undef PACKAGE_MAIL
 
+/* Marker for whether Cygwin is being used */
+#undef IS_CYGWIN
+
+#ifdef IS_CYGWIN
+/* Needed to get some GNU LIBC functions in Cygwin */
+#define _GNU_SOURCE
+#endif
+
+
 #define APT_8_CLEANER_HEADERS
 #define APT_9_CLEANER_HEADERS
 #define APT_10_CLEANER_HEADERS
diff --git a/buildlib/environment.mak.in b/buildlib/environment.mak.in
index b0a8d9d..f68117d 100644
--- a/buildlib/environment.mak.in
+++ b/buildlib/environment.mak.in
@@ -1,3 +1,5 @@
+.DELETE_ON_ERROR:
+
 # This file contains everything that autoconf guessed for your system.
 # if you want you can edit it, just don't re-run configure.
 
diff --git a/buildlib/library.mak b/buildlib/library.mak
index cc0286d..0183184 100644
--- a/buildlib/library.mak
+++ b/buildlib/library.mak
@@ -16,6 +16,22 @@
 # See defaults.mak for information about LOCAL
 
 # Some local definitions
+
+ifeq ($(HOST_OS),cygwin)
+uniq = $(strip $(if $1,$(firstword $1) $(call uniq,$(filter-out $(firstword $1),$1))))
+ifeq ($(HAVE_ZLIB),yes)
+SLIBS += -lz
+endif
+ifeq ($(HAVE_BZ2),yes)
+SLIBS += -lbz2
+endif
+ifeq ($(HAVE_LZMA),yes)
+SLIBS += -llzma
+endif
+SLIBS += -liconv -lintl
+SLIBS := $(call uniq,$(SLIBS))
+endif
+
 LOCAL := lib$(LIBRARY).so.$(MAJOR).$(MINOR)
 $(LOCAL)-OBJS := $(addprefix $(OBJ)/,$(addsuffix .opic,$(notdir $(basename $(SOURCE)))))
 $(LOCAL)-DEP := $(addprefix $(DEP)/,$(addsuffix .opic.d,$(notdir $(basename $(SOURCE)))))
diff --git a/buildlib/program.mak b/buildlib/program.mak
index 71c265f..b3511e4 100644
--- a/buildlib/program.mak
+++ b/buildlib/program.mak
@@ -12,6 +12,22 @@
 # See defaults.mak for information about LOCAL
 
 # Some local definitions
+
+ifeq ($(HOST_OS),cygwin)
+uniq = $(strip $(if $1,$(firstword $1) $(call uniq,$(filter-out $(firstword $1),$1))))
+ifeq ($(HAVE_ZLIB),yes)
+SLIBS += -lz
+endif
+ifeq ($(HAVE_BZ2),yes)
+SLIBS += -lbz2
+endif
+ifeq ($(HAVE_LZMA),yes)
+SLIBS += -llzma
+endif
+SLIBS += -liconv -lintl
+SLIBS := $(call uniq,$(SLIBS))
+endif
+
 LOCAL := $(PROGRAM)
 $(LOCAL)-OBJS := $(addprefix $(OBJ)/,$(addsuffix .o,$(notdir $(basename $(SOURCE)))))
 $(LOCAL)-DEP := $(addprefix $(DEP)/,$(addsuffix .o.d,$(notdir $(basename $(SOURCE)))))
diff --git a/cmdline/makefile b/cmdline/makefile
index b7c35dd..0e54d54 100644
--- a/cmdline/makefile
+++ b/cmdline/makefile
@@ -7,49 +7,49 @@ include ../buildlib/defaults.mak
 
 # The apt program
 PROGRAM=apt
-SLIBS = -lapt-pkg -lapt-private $(INTLLIBS)
+SLIBS = -lapt-private -lapt-pkg $(INTLLIBS)
 LIB_MAKES = apt-pkg/makefile apt-private/makefile
 SOURCE = apt.cc
 include $(PROGRAM_H)
 
 # The apt-cache program
 PROGRAM=apt-cache
-SLIBS = -lapt-pkg -lapt-private $(INTLLIBS)
+SLIBS = -lapt-private -lapt-pkg $(INTLLIBS)
 LIB_MAKES = apt-pkg/makefile apt-private/makefile
 SOURCE = apt-cache.cc
 include $(PROGRAM_H)
 
 # The apt-get program
 PROGRAM=apt-get
-SLIBS = -lapt-pkg -lapt-private $(INTLLIBS)
+SLIBS = -lapt-private -lapt-pkg $(INTLLIBS)
 LIB_MAKES = apt-pkg/makefile apt-private/makefile
-SOURCE = apt-get.cc 
+SOURCE = apt-get.cc
 include $(PROGRAM_H)
 
 # The apt-config program
 PROGRAM=apt-config
-SLIBS = -lapt-pkg -lapt-private $(INTLLIBS)
+SLIBS = -lapt-private -lapt-pkg $(INTLLIBS)
 LIB_MAKES = apt-pkg/makefile apt-private/makefile
 SOURCE = apt-config.cc
 include $(PROGRAM_H)
 
 # The apt-cdrom program
 PROGRAM=apt-cdrom
-SLIBS = -lapt-pkg -lapt-private $(INTLLIBS)
+SLIBS = -lapt-private -lapt-pkg $(INTLLIBS)
 LIB_MAKES = apt-pkg/makefile apt-private/makefile
-SOURCE = apt-cdrom.cc 
+SOURCE = apt-cdrom.cc
 include $(PROGRAM_H)
 
 # The apt-mark program
 PROGRAM=apt-mark
-SLIBS = -lapt-pkg -lapt-private $(INTLLIBS)
+SLIBS = -lapt-private -lapt-pkg $(INTLLIBS)
 LIB_MAKES = apt-pkg/makefile apt-private/makefile
 SOURCE = apt-mark.cc
 include $(PROGRAM_H)
 
 # The apt-helper
 PROGRAM=apt-helper
-SLIBS = -lapt-pkg -lapt-private $(INTLLIBS)
+SLIBS = -lapt-private -lapt-pkg $(INTLLIBS)
 LIB_MAKES = apt-pkg/makefile apt-private/makefile
 SOURCE = apt-helper.cc
 include $(PROGRAM_H)
diff --git a/configure.ac b/configure.ac
index ebffea4..719d4a6 100644
--- a/configure.ac
+++ b/configure.ac
@@ -126,7 +126,7 @@ dnl This is often the dpkg architecture string.
 dnl First check against the full canonical canoncial-system-type in $target
 dnl and if that fails, just look for the cpu
 AC_MSG_CHECKING(debian architecture)
-archset="`dpkg-architecture -qDEB_HOST_ARCH`"
+archset="`dpkg-architecture -qDEB_HOST_ARCH 2>/dev/null`"
 if test "x$archset" = "x"; then
    AC_MSG_ERROR([failed: use --host= or output from dpkg-architecture])
 fi
@@ -198,4 +198,12 @@ fi
 AC_SUBST(USE_NLS)
 AC_PATH_PROG(BASH, bash)
 
+case "$host" in
+   *cygwin*) IS_CYGWIN="yes";;
+esac
+
+if test x"$IS_CYGWIN" = "xyes"; then
+    AC_DEFINE(IS_CYGWIN)
+fi
+
 AC_OUTPUT(environment.mak:buildlib/environment.mak.in makefile:buildlib/makefile.in doc/Doxyfile:doc/Doxyfile.in,make -s dirs)
diff --git a/doc/docbook-html-style.xsl b/doc/docbook-html-style.xsl
index e4af9f5..1cad678 100644
--- a/doc/docbook-html-style.xsl
+++ b/doc/docbook-html-style.xsl
@@ -2,7 +2,7 @@
 <xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="1.0">
 
   <!-- Import our base stylesheet -->
-  <xsl:import href="/usr/share/xml/docbook/stylesheet/docbook-xsl/xhtml-1_1/chunk.xsl" />
+  <xsl:import href="/usr/share/sgml/docbook/xsl-stylesheets/xhtml-1_1/chunk.xsl" />
 
   <!-- Since we use xsltproc (not saxon), add a workaround to ensure UTF-8 -->
   <xsl:template xmlns="http://www.w3.org/1999/xhtml" name="head.content.generator">
diff --git a/doc/docbook-text-style.xsl b/doc/docbook-text-style.xsl
index 376dded..10b60e0 100644
--- a/doc/docbook-text-style.xsl
+++ b/doc/docbook-text-style.xsl
@@ -1,8 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="1.0">
 
-  <xsl:import href="/usr/share/xml/docbook/stylesheet/docbook-xsl/xhtml-1_1/docbook.xsl" />
-
+  <xsl:import href="/usr/share/sgml/docbook/xsl-stylesheets/xhtml-1_1/docbook.xsl" />
   <!-- Parameters for optimal text output. -->
   <xsl:param name="callout.graphics" select="0"/>
   <xsl:param name="callout.unicode" select="0"/>
diff --git a/doc/manpage-style.xsl b/doc/manpage-style.xsl
index a780bad..50659f3 100644
--- a/doc/manpage-style.xsl
+++ b/doc/manpage-style.xsl
@@ -2,8 +2,7 @@
  xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
  version="1.0">
 
-<xsl:import href="/usr/share/xml/docbook/stylesheet/nwalsh/manpages/docbook.xsl" />
-
+<xsl:import href="/usr/share/sgml/docbook/xsl-stylesheets/manpages/docbook.xsl" />
 <xsl:param name="man.output.encoding" select="'UTF-8'" />
 
 <xsl:template match="email">&lt;<xsl:apply-templates/>&gt;</xsl:template>
diff --git a/methods/connect.cc b/methods/connect.cc
index e2cbf4f..3778e4e 100644
--- a/methods/connect.cc
+++ b/methods/connect.cc
@@ -111,7 +111,13 @@ static bool DoConnect(struct addrinfo *Addr,std::string Host,
 
    // Check the socket for an error condition
    unsigned int Err;
+
+#ifdef __CYGWIN__
+   socklen_t Len = sizeof(Err);
+#else
    unsigned int Len = sizeof(Err);
+#endif
+
    if (getsockopt(Fd,SOL_SOCKET,SO_ERROR,&Err,&Len) != 0)
       return _error->Errno("getsockopt",_("Failed"));
    
diff --git a/methods/ftp.cc b/methods/ftp.cc
index 66787a7..05fb741 100644
--- a/methods/ftp.cc
+++ b/methods/ftp.cc
@@ -711,7 +711,11 @@ bool FTPConn::CreateDataFd()
       if (WaitFd(DataFd,true,TimeOut) == false)
 	 return _error->Error(_("Could not connect data socket, connection timed out"));
       unsigned int Err;
+#ifdef __CYGWIN__
+      socklen_t Len = sizeof(Err);
+#else
       unsigned int Len = sizeof(Err);
+#endif
       if (getsockopt(DataFd,SOL_SOCKET,SO_ERROR,&Err,&Len) != 0)
 	 return _error->Errno("getsockopt",_("Failed"));
       if (Err != 0)
diff --git a/methods/makefile b/methods/makefile
index 6b77812..f5d7917 100644
--- a/methods/makefile
+++ b/methods/makefile
@@ -81,7 +81,7 @@ include $(PROGRAM_H)
 
 # The mirror method
 PROGRAM=mirror
-SLIBS = -lapt-pkg $(SOCKETLIBS)
+SLIBS = -lapt-pkg $(SOCKETLIBS) $(INTLLIBS)
 LIB_MAKES = apt-pkg/makefile
 SOURCE = mirror.cc http.cc rfc2553emu.cc connect.cc server.cc
 include $(PROGRAM_H)
diff --git a/test/interactive-helper/makefile b/test/interactive-helper/makefile
index 8dc014b..2a65181 100644
--- a/test/interactive-helper/makefile
+++ b/test/interactive-helper/makefile
@@ -14,14 +14,14 @@ include $(PROGRAM_H)
 
 # Program for testing the tar/deb extractor
 PROGRAM=testdeb
-SLIBS = -lapt-pkg -lapt-inst
+SLIBS = -lapt-inst -lapt-pkg 
 LIB_MAKES = apt-pkg/makefile apt-inst/makefile
 SOURCE = testdeb.cc
 include $(PROGRAM_H)
 
 # Program for testing tar extraction
 PROGRAM=extract-control
-SLIBS = -lapt-pkg -lapt-inst
+SLIBS = -lapt-inst -lapt-pkg 
 LIB_MAKES = apt-pkg/makefile apt-inst/makefile
 SOURCE = extract-control.cc
 include $(PROGRAM_H)
